### DISCUSSION TITLE
Gradient of forces with respect to positions

### DISCUSSION THREAD

**Discussion by sigbjobo** [mir-group/nequip#321] [Q&A] [answered]
Hi, I am working on a project where I need to compute the gradient forces with respect to particle coordinates. Right now, I am thinking to do this numerically, but it is less elegant than doing it through directly.

Is there a simple way to achieve this with the current version of nequip/allegro? And if not, is there a simple way to implement this?

**Comment by Linux-cpp-lisp** [comment, marked as answer]
Hi @sigbjobo ,

I think this should be addressed by https://github.com/mir-group/nequip/discussions/292#discussioncomment-4702180.  Please note that since that post the `functorch` / `torch.autograd.functional` APIs have evolved and matured somewhat, and there may be better / faster ways to do it, such as trying with `vectorize=True`.

I'd still be happy to accept a PR adding this and don't think it should be too much work.  (Fair warning, 2nd derivatives in PyTorch were, at least last time I checked, not particularly fast.)

**Reply by sigbjobo** [reply]
Thanks, this sounds promising! I will definitely submit a PR if we manage to do it.

**Reply by Linux-cpp-lisp** [reply]
Great! Yes, it should be fairly straightforward I hope.
